src/DotNetty.Transport/Channels/Sockets/SocketDatagramChannel.cs

```diff
@@ -0,0 +1,479 @@
+#else
+            if (ExecutionContext.IsFlowSuppressed())
+            {
+                pending = this.Socket.ReceiveFromAsync(operation);
+            }
+            else
+            {
+                using (ExecutionContext.SuppressFlow())
+                {
+                    pending = this.Socket.ReceiveFromAsync(operation);
+                }
+            }
+#endif
+            if (!pending)
+            {
+                this.EventLoop.Execute(ReceiveFromCompletedSyncCallback, this.Unsafe, operation);
+            }
+        }
+
+        protected override int DoReadMessages(List<object> buf)
+        {
+            Contract.Requires(buf != null);
+
+            SocketChannelAsyncOperation operation = this.ReadOperation;
+
+            var data = (IByteBuffer)operation.UserToken;
+            bool free = true;
+            try
+            {
+                int received = operation.BytesTransferred;
+                if (received <= 0)
+                {
+                    return 0;
+                }
+
+                IRecvByteBufAllocatorHandle handle = this.Unsafe.RecvBufAllocHandle;
+                handle.LastBytesRead = received;
+                data.SetWriterIndex(data.WriterIndex + received);
+                EndPoint remoteAddress = operation.RemoteEndPoint;
+                buf.Add(new DatagramPacket(data, remoteAddress, this.LocalAddress));
+                free = false;
+
+                return 1;
+            }
+            catch (Exception exception)
+            {
+                Logger.Error(exception);
+                return -1;
+            }
+            finally
+            {
+                if (free)
+                {
+                    data.Release();
+                }
+
+                operation.UserToken = null;
+            }
+        }
+
+        static void OnReceiveFromCompletedSync(object u, object p) => ((ISocketChannelUnsafe)u).FinishRead((SocketChannelAsyncOperation)p);
+
+        protected override void ScheduleMessageWrite(object message)
+        {
+            var envelope = message as IAddressedEnvelope<IByteBuffer>;
+            if (envelope == null)
+            {
+                throw new InvalidOperationException(
+                    $"Unexpected type: {message.GetType().FullName}, expecting DatagramPacket or IAddressedEnvelope.");
+            }
+
+            IByteBuffer data = envelope.Content;
+            int length = data.ReadableBytes;
+            if (length == 0)
+            {
+                return;
+            }
+
+            SocketChannelAsyncOperation operation = this.PrepareWriteOperation(data.GetIoBuffer(data.ReaderIndex, length));
+            operation.RemoteEndPoint = envelope.Recipient;
+            this.SetState(StateFlags.WriteScheduled);
+            bool pending = this.Socket.SendToAsync(operation);
+            if (!pending)
+            {
+               ((ISocketChannelUnsafe)this.Unsafe).FinishWrite(operation);
```

Comment:
I've checked with .NET Core developers and it's the same way on Linux as of today but I'll post a fix to .NET Core so it will become a concern going forward.